### PR TITLE
enhancement of compare feature

### PR DESCRIPTION
**Pull request by a1012112796**
1. add three compare modes instead of single mode, (that's like gitub), they are:

TIP: Compare mode was deteched by the compare link format.

a: `Compare in same repository`: will only show branch/tags dropdown, wich used to compare or create pull request in same repo.

image

image

b: `Compare across repositorys` which will not only show branch/tags dropdown, but also show repository dropdown, by default will single show fork repository of this repo in this list, but will load more repositorys if you entry a word in filter box.

image

image

c: `Compare with repository in other service`, a new mode which github not suport also :), in this mode, the head repo select box will be replaced with a repo url entry box, you can entry a git repo link to other website, then you can comare with it.

image

- [ ] TODO: will suport creating pull request from repository in other service in future, now only support compare
- note: To prevent to much time cost, has add `--depth=100` as a limit.

3. show a single error log instead of a `404 page` when the head repo or user, the head or base ref
is not exist. then user still can choose other option easyly.

image

image

### CODE REVIEW

**Comment by delvh** [comment]
Regarding `c)`:
What about calling it `Compare with external repo(sitory)` instead?

**Comment by lunny** [comment]
1 and 2 could be merged?

File: models/perm/access/repo_permission.go [hunk 1]
```diff
@@ -128,6 +128,11 @@ func (p *Permission) LogString() string {
 
 // GetUserRepoPermission returns the user permissions to the repository
 func GetUserRepoPermission(ctx context.Context, repo *repo_model.Repository, user *user_model.User) (perm Permission, err error) {
+	if repo.ID == -1 {
```
Comment (lunny) [inline models/perm/access/repo_permission.go]: Use GhostID

Comment (lunny) [reply models/perm/access/repo_permission.go]: And it depends on whether the repo is public or not

**Pull request closed without merging**